File: Suduku.py
class Sudoku:
    GRID_SIZE = 9
    BOX_SIZE = 3

    def __init__(self, input_str):
        """
        对象创建和初始化。
        将输入字符串解析为九宫格的初始状态。
        """
        self.grid = [[0] * Sudoku.GRID_SIZE for _ in range(Sudoku.GRID_SIZE)]
        index = 0
        for row in range(Sudoku.GRID_SIZE):
            for col in range(Sudoku.GRID_SIZE):
                self.grid[row][col] = int(input_str[index])
                index += 1

    def clone(self):
        """
        克隆方法，创建当前九宫格的副本。
        """
        cloned_grid = Sudoku('0' * Sudoku.GRID_SIZE * Sudoku.GRID_SIZE)
        for row in range(Sudoku.GRID_SIZE):
            for col in range(Sudoku.GRID_SIZE):
                cloned_grid.grid[row][col] = self.grid[row][col]
        return cloned_grid

    def serialize(self):
        """
        串行化方法，将九宫格状态转换为字符串。
        """
        serialized_str = ''
        for row in range(Sudoku.GRID_SIZE):
            for col in range(Sudoku.GRID_SIZE):
                serialized_str += str(self.grid[row][col])
        return serialized_str

    def __eq__(self, other):
        """
        比较方法，判断两个九宫格是否相等。
        """
        if not isinstance(other, Sudoku):
            return False
        for row in range(Sudoku.GRID_SIZE):
            for col in range(Sudoku.GRID_SIZE):
                if self.grid[row][col]!= other.grid[row][col]:
                    return False
        return True

    def get_candidates(self, row, col):
        """
        获取指定单元格的候选值。
        """
        candidates = set(range(1, Sudoku.GRID_SIZE + 1))
        # 检查所在行
        for i in range(Sudoku.GRID_SIZE):
            if self.grid[row][i] in candidates:
                candidates.remove(self.grid[row][i])
        # 检查所在列
        for i in range(Sudoku.GRID_SIZE):
            if self.grid[i][col] in candidates:
                candidates.remove(self.grid[i][col])
        # 检查所在 3x3 子格
        start_row, start_col = Sudoku.BOX_SIZE * (row // Sudoku.BOX_SIZE), Sudoku.BOX_SIZE * (col // Sudoku.BOX_SIZE)
        for i in range(start_row, start_row + Sudoku.BOX_SIZE):
            for j in range(start_col, start_col + Sudoku.BOX_SIZE):
                if self.grid[i][j] in candidates:
                    candidates.remove(self.grid[i][j])
        return candidates

    def getRow(self, row_index):
        """
        获取指定行。
        """
        return self.grid[row_index]

    def getColumn(self, col_index):
        """
        获取指定列。
        """
        return [self.grid[i][col_index] for i in range(Sudoku.GRID_SIZE)]

    def getBox(self, row, col):
        """
        获取指定行和列所在的 3x3 子格。
        """
        box_row_start = Sudoku.BOX_SIZE * (row // Sudoku.BOX_SIZE)
        box_col_start = Sudoku.BOX_SIZE * (col // Sudoku.BOX_SIZE)
        box = []
        for i in range(box_row_start, box_row_start + Sudoku.BOX_SIZE):
            for j in range(box_col_start, box_col_start + Sudoku.BOX_SIZE):
                box.append(self.grid[i][j])
        return box

    def solve(self):
        """
        使用回溯算法尝试求解数独
        """
        for row in range(self.GRID_SIZE):
            for col in range(self.GRID_SIZE):
                if self.grid[row][col] == 0:
                    for num in range(1, self.GRID_SIZE + 1):
                        if self.is_valid_move(row, col, num):
                            self.grid[row][col] = num
                            if self.solve():
                                return True
                            else:
                                self.grid[row][col] = 0
                    return False
        return True

    def is_solved(self):
        """
        判断数独问题是否求解成功
        """
        for row in range(self.GRID_SIZE):
            for col in range(self.GRID_SIZE):
                if self.grid[row][col] == 0:
                    return False
                if not self.is_valid_move(row, col, self.grid[row][col]):
                    return False
        return True

    def print_grid(self):
        """
        打印数独
        """
        for row in self.grid:
            print(' '.join(map(str, row)))

    def count_empty_cells(self):
        """
        统计数独中还有多少未填格
        """
        count = 0
        for row in range(self.GRID_SIZE):
            for col in range(self.GRID_SIZE):
                if self.grid[row][col] == 0:
                    count += 1
        return count
def test_sudoku():
    """
    测试代码和测试用例。
    """
    input_str = '017903600000080000900000507072010430000402070064370250701000065000030000005601720'
    sudoku = Sudoku(input_str)
    assert len(sudoku.grid) == Sudoku.GRID_SIZE
    assert len(sudoku.grid[0]) == Sudoku.GRID_SIZE
    cloned_sudoku = sudoku.clone()
    assert sudoku == cloned_sudoku
    serialized_str = sudoku.serialize()
    assert serialized_str == input_str
    candidates = sudoku.get_candidates(0, 0)
    assert isinstance(candidates, set)
    row = sudoku.getRow(0)
    assert len(row) == Sudoku.GRID_SIZE
    column = sudoku.getColumn(0)
    assert len(column) == Sudoku.GRID_SIZE
    box = sudoku.getBox(0, 0)
    assert len(box) == Sudoku.BOX_SIZE * Sudoku.BOX_SIZE
    print("All tests passed!")

if __name__ == '__main__':
    test_sudoku()

File: Suduku.java
class Sudoku {
    public static final int GRID_SIZE = 9;
    public static final int BOX_SIZE = 3;
    private int[][] grid;

    public Sudoku(String inputStr) {
        grid = new int[GRID_SIZE][GRID_SIZE];
        int index = 0;
        for (int row = 0; row < GRID_SIZE; row++) {
            for (int col = 0; col < GRID_SIZE; col++) {
                grid[row][col] = inputStr.charAt(index) - '0';
                index++;
            }
        }
    }

    public Sudoku clone() {
        Sudoku clonedSudoku = new Sudoku("0".repeat(GRID_SIZE * GRID_SIZE));
        for (int row = 0; row < GRID_SIZE; row++) {
            for (int col = 0; col < GRID_SIZE; col++) {
                clonedSudoku.grid[row][col] = grid[row][col];
            }
        }
        return clonedSudoku;
    }
    //串行化
    public String serialize() {
        StringBuilder serializedStr = new StringBuilder();
        for (int row = 0; row < GRID_SIZE; row++) {
            for (int col = 0; col < GRID_SIZE; col++) {
                serializedStr.append(grid[row][col]);
            }
        }
        return serializedStr.toString();
    }

    public boolean equals(Object other) {
        if (!(other instanceof Sudoku)) {
            return false;
        }
        Sudoku otherSudoku = (Sudoku) other;
        for (int row = 0; row < GRID_SIZE; row++) {
            for (int col = 0; col < GRID_SIZE; col++) {
                if (grid[row][col]!= otherSudoku.grid[row][col]) {
                    return false;
                }
            }
        }
        return true;
    }

    public int[] getRow(int rowIndex) {
        return grid[rowIndex];
    }

    public int[] getColumn(int colIndex) {
        int[] column = new int[GRID_SIZE];
        for (int row = 0; row < GRID_SIZE; row++) {
            column[row] = grid[row][colIndex];
        }
        return column;
    }

    public int[] getBox(int row, int col) {
        int[] box = new int[BOX_SIZE * BOX_SIZE];
        int boxRowStart = BOX_SIZE * (row / BOX_SIZE);
        int boxColStart = BOX_SIZE * (col / BOX_SIZE);
        int index = 0;
        for (int i = boxRowStart; i < boxRowStart + BOX_SIZE; i++) {
            for (int j = boxColStart; j < boxColStart + BOX_SIZE; j++) {
                box[index++] = grid[i][j];
            }
        }
        return box;
    }
    //判断（row、col）格是否能填入num
    public boolean is_valid_move(int row, int col, int num) {
        // 检查行
        for (int i = 0; i < GRID_SIZE; i++) {
            if (grid[row][i] == num) {
                return false;
            }
        }
        // 检查列
        for (int i = 0; i < GRID_SIZE; i++) {
            if (grid[i][col] == num) {
                return false;
            }
        }
        // 检查 3x3 子格
        int boxRowStart = BOX_SIZE * (row / BOX_SIZE);
        int boxColStart = BOX_SIZE * (col / BOX_SIZE);
        for (int i = boxRowStart; i < boxRowStart + BOX_SIZE; i++) {
            for (int j = boxColStart; j < boxColStart + BOX_SIZE; j++) {
                if (grid[i][j] == num) {
                    return false;
                }
            }
        }
        return true;
    }
    //使用回溯方法尝试解决该数独问题
    public boolean solve() {
        for (int row = 0; row < GRID_SIZE; row++) {
            for (int col = 0; col < GRID_SIZE; col++) {
                if (grid[row][col] == 0) {
                    for (int num = 1; num <= GRID_SIZE; num++) {
                        if (is_valid_move(row, col, num)) {
                            grid[row][col] = num;
                            if (solve()) {
                                return true;
                            } else {
                                grid[row][col] = 0;
                            }
                        }
                    }
                    return false;
                }
            }
        }
        return true;
    }
    //判断数独问题是否被解决
    public boolean is_solved() {
        for (int row = 0; row < GRID_SIZE; row++) {
            for (int col = 0; col < GRID_SIZE; col++) {
                if (grid[row][col] == 0) {
                    return false;
                }
                if (!is_valid_move(row, col, grid[row][col])) {
                    return false;
                }
            }
        }
        return true;
    }
    //打印数独表格
    public void print_grid() {
        for (int[] row : grid) {
            for (int num : row) {
                System.out.print(num + " ");
            }
            System.out.println();
        }
    }
    //统计还剩多少未填格子
    public int count_empty_cells() {
        int count = 0;
        for (int row = 0; row < GRID_SIZE; row++) {
            for (int col = 0; col < GRID_SIZE; col++) {
                if (grid[row][col] == 0) {
                    count++;
                }
            }
        }
        return count;
    }
}

class SudokuTest {
    public static void main(String[] args) {
        // 测试构造函数和序列化
        String inputStr = "017903600000080000900000507072010430000402070064370250701000065000030000005601720";
        Sudoku sudoku = new Sudoku(inputStr);
        String serializedStr = sudoku.serialize();
        assert serializedStr.equals(inputStr) : "序列化错误";

        // 测试克隆
        Sudoku clonedSudoku = sudoku.clone();
        assert sudoku.equals(clonedSudoku) : "克隆错误";

        // 测试获取行、列和子格
        int[] row = sudoku.getRow(0);
        assert row.length == Sudoku.GRID_SIZE : "获取行错误";
        int[] column = sudoku.getColumn(0);
        assert column.length == Sudoku.GRID_SIZE : "获取列错误";
        int[] box = sudoku.getBox(0, 0);
        assert box.length == Sudoku.BOX_SIZE * Sudoku.BOX_SIZE : "获取子格错误";

        // 测试有效性检查
        assert sudoku.is_valid_move(0, 0, 1) : "有效性检查错误";
        assert!sudoku.is_valid_move(0, 0, row[0]) : "有效性检查错误";

        // 测试求解和是否已解决
        boolean solvedBefore = sudoku.is_solved();
        assert!solvedBefore : "初始状态错误地被认为已解决";
        sudoku.solve();
        boolean solvedAfter = sudoku.is_solved();
        assert solvedAfter : "求解错误";

        // 测试打印
        System.out.println("原始数独：");
        new Sudoku(inputStr).print_grid();
        System.out.println("求解后的数独：");
        sudoku.print_grid();

        // 测试统计空白单元格
        int emptyCellsBefore = new Sudoku(inputStr).count_empty_cells();
        assert emptyCellsBefore > 0 : "空白单元格统计错误";
        int emptyCellsAfter = sudoku.count_empty_cells();
        assert emptyCellsAfter == 0 : "空白单元格统计错误";
    }
}
